Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs

```diff
@@ -310,6 +317,36 @@ public void UnregisterBackgroundProgram(EntityUid loaderUid, EntityUid cartridge
 
     protected override void OnItemInserted(EntityUid uid, CartridgeLoaderComponent loader, EntInsertedIntoContainerMessage args)
     {
+        if (args.Container.ID == PdaComponent.PdaIdSlotId)
+        {
+            //_sawmill.Debug("Id insertion detected");
+            if ((TryComp(uid, out ContainerManagerComponent? cont)))
+            {
+                //_sawmill.Debug("Trying program");
+                if (TryGetProgram<MessagesCartridgeComponent>(uid, out var programUid, out var messagesCartridgeComponent, false, loader, cont))
+                {
+                    //_sawmill.Debug("program found");
+                    if ((TryComp(uid, out PdaComponent? pdaComponent)) && (programUid is EntityUid realProgramUid))
+                    {
+                        //_sawmill.Debug("pda found");
+                        if (TryComp(args.Entity, out IdCardComponent? idCardComponent))
+                        {
+                            //_sawmill.Debug("found id");
+                            //_sawmill.Debug("updating cart data");
+                            messagesCartridgeComponent.ConnectedId = args.Entity;
+                            messagesCartridgeComponent.UserUid = args.Entity.ToString();
+                            //_sawmill.Debug("data updated, looking for server");
+                            if (_messagesCartridgeSystem.GetActiveServer(messagesCartridgeComponent,Transform(uid).MapID) is var (_, server))
+                            {
+                                //_sawmill.Debug("server found");
+                                _messagesCartridgeSystem.PullFromServer(realProgramUid, messagesCartridgeComponent, server);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
         if (args.Container.ID != InstalledContainerId && args.Container.ID != loader.CartridgeSlot.ID)
```

Comment:
My issue here is I want the cartridge to react to the id being inserted into the PDA.
From what I understand the local event raised in that case reaches the cartridgeloader, but since the cartridge itself seems to be a separate entity, I don't think it would receive the event.
One way I could see doing this without hardcoding the cartridge would be to have a generic "ID was inserted" function in the cartridges that I would then call for each cartridge.

Reply:
five levels of indention is insane

Reply:
> My issue here is I want the cartridge to react to the id being inserted into the PDA. From what I understand the local event raised in that case reaches the cartridgeloader, but since the cartridge itself seems to be a separate entity, I don't think it would receive the event. One way I could see doing this without hardcoding the cartridge would be to have a generic "ID was inserted" function in the cartridges that I would then call for each cartridge.

Add a new event to cartidges that get loaded or something, this indentation makes the entire thing hard to manage and prevent bugs.